src/util/structured_data.cpp

```diff
@@ -0,0 +1,218 @@
+// Copyright 2016-2020 Diffblue Limited. All Rights Reserved.
+
+#include "structured_data.h"
+#include "range.h"
+#include "string_utils.h"
+#include <algorithm>
+
+labelt::labelt(std::vector<std::string> components)
+{
+  auto to_lower_string = [](const std::string &s) -> std::string {
+    std::string lower_case = s;
+    std::for_each(
+      lower_case.begin(), lower_case.end(), [](char &c) { c = ::tolower(c); });
+    return lower_case;
+  };
+  this->components = make_range(components)
+                       .map(to_lower_string)
+                       .collect<std::vector<std::string>>();
+}
```

Comment:
Could we make it a precondition that the inputs need to be in all lowercase? That would ensure better consistency in the uses of the API. Additionally we could add a precondition checking that there is at least one component (then we can get rid of the `!components.empty()` checks), and a precondition checking that each component is non-empty.